frame/support/src/traits/tokens/fungibles/balanced.rs

```diff
@@ -0,0 +1,367 @@
+		let credit = match Self::withdraw(asset, who, amount) {
+			Err(_) => return Err(debt),
+			Ok(d) => d,
+		};
+		match credit.offset(debt) {
+			Ok(SameOrOther::None) => Ok(CreditOf::<AccountId, Self>::zero(asset)),
+			Ok(SameOrOther::Same(dust)) => Ok(dust),
+			Ok(SameOrOther::Other(rest)) => {
+				debug_assert!(false, "ok withdraw return must be at least debt value; qed");
+				Err(rest)
+			}
+			Err(_) => {
+				debug_assert!(false, "debt.asset is credit.asset; qed");
+				Ok(CreditOf::<AccountId, Self>::zero(asset))
+			}
+		}
+	}
+}
+
+pub trait Unbalanced<AccountId>: Inspect<AccountId> {
+	/// Set the `asset` balance of `who` to `amount`. If this cannot be done for some reason (e.g.
+	/// because the account cannot be created or an overflow) then an `Err` is returned.
+	fn set_balance(asset: Self::AssetId, who: &AccountId, amount: Self::Balance) -> DispatchResult;
+
+	/// Set the total issuance of `asset` to `amount`.
+	fn set_total_issuance(asset: Self::AssetId, amount: Self::Balance);
+
+	/// Reduce the `asset` balance of `who` by `amount`. If it cannot be reduced by that amount for
+	/// some reason, return `Err` and don't reduce it at all. If Ok, return the imbalance.
+	///
+	/// Minimum balance will be respected and the returned imbalance may be up to
+	/// `Self::minimum_balance() - 1` greater than `amount`.
+	fn decrease_balance(asset: Self::AssetId, who: &AccountId, amount: Self::Balance)
+		-> Result<Self::Balance, DispatchError>
+	{
+		let old_balance = Self::balance(asset, who);
+		let (mut new_balance, mut amount) = if old_balance < amount {
+			Err(TokenError::NoFunds)?
+		} else {
+			(old_balance - amount, amount)
+		};
+		if new_balance < Self::minimum_balance(asset) {
+			amount = amount.saturating_add(new_balance);
+			new_balance = Zero::zero();
+		}
+		// Defensive only - this should not fail now.
+		Self::set_balance(asset, who, new_balance)?;
+		Ok(amount)
+	}
+
+	/// Reduce the `asset` balance of `who` by the most that is possible, up to `amount`.
+	///
+	/// Minimum balance will be respected and the returned imbalance may be up to
+	/// `Self::minimum_balance() - 1` greater than `amount`.
+	///
+	/// Return the imbalance by which the account was reduced.
+	fn decrease_balance_at_most(asset: Self::AssetId, who: &AccountId, amount: Self::Balance)
+		-> Self::Balance
+	{
+		let old_balance = Self::balance(asset, who);
+		let (mut new_balance, mut amount) = if old_balance < amount {
+			(Zero::zero(), old_balance)
+		} else {
+			(old_balance - amount, amount)
+		};
+		let minimum_balance = Self::minimum_balance(asset);
+		if new_balance < minimum_balance {
+			amount = amount.saturating_add(new_balance);
+			new_balance = Zero::zero();
+		}
+		let mut r = Self::set_balance(asset, who, new_balance);
+		if r.is_err() {
+			// Some error, probably because we tried to destroy an account which cannot be destroyed.
+			if amount > minimum_balance {
+				new_balance += minimum_balance;
+				amount -= minimum_balance;
```

Comment:
this isnt super accurate right?

ideally you would only add to `new_balance` the amount you need to get to `minimum_balance`

```
amount -= (minimum_balance - new_balance);
new_balance = minimum_balance;

Reply:
```suggestion
				amount -= (minimum_balance - new_balance);
				new_balance = minimum_balance;
```